AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -1491,13 +1472,114 @@ - (void)measureNodeWithBoundsIfNecessary:(CGRect)bounds
+      ASLayoutElementSetCurrentContext(context);
+      didOverrideExistingContext = YES;
+    }
+  }
+  
+  // Figure out previos and pending layout for layout transition
+  auto previousLayout = _calculatedDisplayNodeLayout;
+  auto pendingLayout = [=]() -> std::shared_ptr<ASDisplayNodeLayout> {
+    // Check if  the pending display node layout can be used to transition to
+    if (_pendingDisplayNodeLayout != nullptr &&
+        _pendingDisplayNodeLayout->isDirty() == NO &&
+        CGSizeEqualToSize(_pendingDisplayNodeLayout->layout.size, ASCeilSizeValues(bounds.size))) {
+      // We assume the _pendingDisplayNodeLayout was created by layoutThatFits: to set the size of the node
+      // now it's time to apply it and to become the _calculatedDisplayNodeLayout
+      return _pendingDisplayNodeLayout;
+    }
+  
+    // The _pendingDisplayNodeLayout can not be used to be applied we need to figure out we need to calculate a new
+    // layout based on a constrainedSize
+  
+    // Use as default constrained size the bounds
+    CGSize parentSize = bounds.size;
+    ASSizeRange constrainedSize = [self constrainedSizeForLayoutPass];
+    
+    return std::make_shared<ASDisplayNodeLayout>(
+      [self calculateLayoutThatFits:constrainedSize restrictedToSize:self.style.size relativeToParentSize:parentSize],
+      constrainedSize,
+      parentSize
+    );
+  }();
+  
+  // If the size of the new layout to apply did change from the current bounds, invalidate the whole tree up
+  // so the root node can handle a resizing if necessary
+  if (CGSizeEqualToSize(ASCeilSizeValues(bounds.size), pendingLayout->layout.size) == NO) {
+    [self invalidateSize];
+  }
+
+  if (didCreateNewContext) {
+    ASLayoutElementClearCurrentContext();
+  } else if (didOverrideExistingContext) {
+    context.needsVisualizeNode = !context.needsVisualizeNode;
+    ASLayoutElementSetCurrentContext(context);
+  }
+
+  // Finally transition to pendingLayout
+  ASDisplayNodeAssertNotNil(pendingLayout->layout, @"pendingLayout->layout should not be nil! %@", self);
+
+  _pendingLayoutTransition = [[ASLayoutTransition alloc] initWithNode:self
+                                                        pendingLayout:pendingLayout
+                                                       previousLayout:previousLayout];
+
+  // Only complete the pending layout transition if the node is not a subnode of a node that is currently
+  // in a layout transition
+  if (ASHierarchyStateIncludesLayoutPending(_hierarchyState) == NO) {
+    // Complete the pending layout transition immediately
+    [self _completePendingLayoutTransition];
+  }
+}
+
+- (ASSizeRange)constrainedSizeForLayoutPass
+{
+  CGRect bounds = self.threadSafeBounds;
+  
+  // Checkout if constrained size of pending or calculated display node layout can be used
+  if (_pendingDisplayNodeLayout != nullptr && CGSizeEqualToSize(_pendingDisplayNodeLayout->layout.size, ASCeilSizeValues(bounds.size))) {
```

Comment:
This check seems incorrect to me:

- If a layout is _pending, then by definition, it has *not been applied*.
- If a layout has *not been applied* then there's no reason that the bounds.size should match the layout.size yet.
- Even though it hasn't been applied, the constrainedSize used to calculate the _pendingLayout is the *most recent information* given from the parent about the constrainedSize.  If the constrainedSize has changed, the parent would've notified us, so we can use it directly.

So it seems to me this check should be ONLY for != nullptr, and unconditionally use the constrainedSize if it exists.

I think this will come up if a node has had a recent layoutThatFits: call with NEW constrainedSize, but the layout pass hasn't occurred yet to apply it (and bubble up to parents).  Then setNeedsLayout occurs, and the old constrainedSize is used...